File: go/array/14_longest_common_prefix.go
// Time Complexity: O(n)
// Space Complexity: O(n)
package main

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	for i, letter := range strs[0] {
		for _, word := range strs {
			if i >= len(word) || rune(word[i]) != letter {
				return strs[0][:i]
			}
		}
	}
	return strs[0]
}

File: python/array/14_longest_common_prefix.py
# Time Complexity: O(n)
# Space Complexity: O(n)

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if len(strs) == 0:
            return ""
        
        for i, letter in enumerate(strs[0]):
            for word in strs:
                if i >= len(word) or word[i] != letter:
                    return strs[0][:i]
        return strs[0]